### PR TITLE
feat(REST): new includeAttachments param for search2 API

### PR DESCRIPTION
**Pull request by edalex-ian**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This adds a new param to the search2 endpoint that allows for attachments to be excluded from the result. Doing so can skip the (sometimes) intensive processing required for attachments and thereby speed up responses.

The intention is to use this in the front-end to exclude attachments on search, but then if a user expands the attachments for an item to use the same endpoint but with a `must` of the item `uuid`. To ensure this can be done and still have the UI display how many attachments an item have there is also a new field in the response called `attachmentCount`.

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchResource.scala [hunk 1]
```diff
@@ -104,7 +104,7 @@ class SearchResource {
       searchResults.getOffset,
       searchResults.getCount,
       searchResults.getAvailable,
-      items.map(convertToItem),
+      items.map(i => convertToItem(i, params.includeAttachments)),
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchResource.scala:107]: How about this? 

```
convertToItem(_, params.includeAttachments)
```

**Review by PenghaiZhang — APPROVED** [review]
LGTM! 

Do we need to update the Jest test as well ?

**Comment by edalex-ian** [comment]
> LGTM!
> 
> Do we need to update the Jest test as well ?

Nope, that's in the next PR for the rest-module. :wink:

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchResource.scala (hunk 1 above)
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchResource.scala]: Excellent, thanks @PenghaiZhang . I'm not that familiar with that syntax yet.

**Comment by edalex-ian** [comment]
Thanks @PenghaiZhang , will merge on green.

**Comment by edalex-ian** [comment]
Sorry @PenghaiZhang , could you please do a quick re-review?

Turns out there was a case there (highlighted by the failing `NewSearchPageTest.searchWithLessACLS`) where `getAttachments` can return null. (Under low ACLs it returns null it seems, where as for when there's no attachments - but a user has permissions - it returns an empty collection.)

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala [hunk 2]
```diff
@@ -268,7 +268,7 @@ object SearchHelper {
       collectionId = bean.getCollection.getUuid,
       commentCount = getItemCommentCount(key),
       starRatings = bean.getRating,
-      attachmentCount = bean.getAttachments.size(),
+      attachmentCount = Option(bean.getAttachments).map(_.size()).getOrElse(0),
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:271]: Can just do `_.size` but let's leave it as it's.

**Review by PenghaiZhang — APPROVED** [review]

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala (hunk 2 above)
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala]: Gah, of course! I'll do that me thinks. :wink:

**Pull request merged**